### PR TITLE
691 celts sponsored program

### LINKED ISSUE #691: Make CELTS sponsored events into a program
In order to handle permissions for events that don't have a program we need to change the "other" category to a "CELTS-Sponsored" program. This means all of the places that assume there are events with no program need to be altered.

### PR DESCRIPTION
**Pull request by ojmakinde**
Issue #691

Issue: CELTS-Sponsored Events and All Volunteer Trainings used to have a null value in the program_id column.

Solution: We created a program for CELTS-Sponsored Events and All Volunteer Trainings. Additionally, we ensured all test data were correctly working with our added changes. We changed the links on create event to create events that links to the "Other CELTS-Sponsored Program"

To test, go to Create Event and create a CELTS-Sponsored Event and an All Volunteer Training, both in the past. Add Zach Neill as a volunteer to both events. Visit the database and ensure the both events have a program id value of 9. Then, visit Zach Neill's profile and view his transcript and verify the hours he earned from volunteering at the created events appear and are under the title "CELTS Sponsored Events".

Resolves #691

### CODE REVIEW

File: app/logic/userManagement.py [hunk 1]
```diff
@@ -53,7 +53,8 @@ def changeProgramInfo(newProgramName, newContactEmail, newContactName, newLocati
 def getAllowedPrograms(currentUser):
     """Returns a list of all visible programs depending on who the current user is."""
     if currentUser.isCeltsAdmin:
-        return Program.select().order_by(Program.programName)
+        # Beans: ask if this where clause is an acceptable way to filter out the celts sponsored event from appearing in the list of programs
+        return Program.select().where(Program.programName != "CELTS-Sponsored Event").order_by(Program.programName)
```
Comment (BrianRamsay) [inline app/logic/userManagement.py]: We can just change this to use the id 9, right? Also, why are we not letting admins see these?

**Review by BrianRamsay — CHANGES_REQUESTED** [review]
Instead of having the magic number 9 everywhere, let's define a class variable Program.OTHER_CELTS or something that we can use. (see Certification.BONNER)

File: app/logic/userManagement.py (hunk 1 above)
Comment (hoerstl) [reply app/logic/userManagement.py]: Thanks for the suggestion about the class variable. We are not letting admins see these because the buttons that now create other celts sponsored events is returned from getAllowedTemplates. If we allowed this to return the celts sponsored events as a program, that button would appear twice on the page. Since this function is only used to, from the docstring, "Returns a list of all visible programs", we didn't see a need to return this as a program. This is the only use of this function.

File: app/logic/transcript.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
 def getOtherEventsTranscript(username):
     """
-    Returns a Other Events query object containing all the non-program
+    Returns a Other Events query object containing all the other-program
```
Comment (zachneill) [inline app/logic/transcript.py]: while you're at it changing this docstring, it's "Returns **_an_** Other Events query object..."

File: database/test_data.py [hunk 3]
```diff
@@ -450,6 +450,16 @@
         "contactEmail": "",
         "contactName": ""
     },
+    {
+        "id": 9,
+        "programName": "CELTS-Sponsored Event",
+        "programUrl": "https://www.berea.edu/celts/",
```
Comment (zachneill) [inline database/test_data.py]: dead link, it's actually https://www.berea.edu/centers/center-for-excellence-in-learning-through-service

you're leading your own poor staff and students astray. thats bad and you should feel bad

**Review by zachneill — CHANGES_REQUESTED** [review]
this is fun

File: app/models/program.py [hunk 4]
```diff
@@ -14,6 +14,8 @@ class Program(baseModel):
     contactEmail = CharField(null=True,default='')
     defaultLocation = CharField(null=True,default='')
 
+    OTHERCELTSSPONSORED = 9
```
Comment (AndersonStettner) [inline app/models/program.py]: To keep with how we are already doing this for other things this should be changed to a program field.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

File: database/test_data.py (hunk 3 above)
Comment (hoerstl) [reply database/test_data.py]: Atoning for sins is a blessing

File: app/models/program.py (hunk 4 above)
Comment (hoerstl) [reply app/models/program.py]: That's not quite a field in the database. That variable is simply meant as a static and constant class variable for the Program class that we can use to refer to the magic number 9. I don't believe we want to make it a program field because that would add it to the database. See line 8 in /app/models/certification.py for an example of how we use it for BONNER inside of Certification.

Comment (hoerstl) [reply app/models/program.py]: Okay, i do it

File: tests/code/test_scripts.py [hunk 5]
```diff
@@ -2,6 +2,7 @@
 from datetime import date, datetime, timedelta
 
 from app.models import mainDB
+from app.models.program import Program
```
Comment (AndersonStettner) [inline tests/code/test_scripts.py]: Why did you add this import?

File: app/models/event.py [hunk 6]
```diff
@@ -21,7 +21,7 @@ class Event(baseModel):
     recurringId = IntegerField(null=True)
     contactEmail = CharField(null=True)
     contactName = CharField(null=True)
-    program = ForeignKeyField(Program, null= True)
+    program = ForeignKeyField(Program, default=9)
```
Comment (AndersonStettner) [inline app/models/event.py]: We do not need any default. It can just be `ForeignKeyField(Program)`

File: app/logic/transcript.py [hunk 7]
```diff
@@ -36,7 +36,8 @@ def getProgramTranscript(username):
     EventData = (Event
         .select(Event, fn.SUM(EventParticipant.hoursEarned).alias("hoursEarned"))
         .join(EventParticipant)
-        .where(EventParticipant.user == username)
+        .where(EventParticipant.user == username, 
+        Event.program != 9)
```
Comment (AndersonStettner) [inline app/logic/transcript.py]: The where clause is short enough to all be on the same line, also we should not be using a magic number.

File: app/logic/transcript.py [hunk 8]
```diff
@@ -20,7 +20,7 @@ def getOtherEventsTranscript(username):
                     .join(Program, JOIN.LEFT_OUTER).switch(Event)
                     .join(Term)
                     .where(EventParticipant.user==username,
-                    Event.program == None)
+                    Event.program == 9)
```
Comment (AndersonStettner) [inline app/logic/transcript.py]: This where clause is short enough to all be on the same line, also we should not be using a magic number.

File: app/logic/events.py [hunk 9]
```diff
@@ -227,7 +227,7 @@ def getOtherEvents(term):
                             .where(Event.term == term,
                                    Event.isTraining == False,
                                    Event.isAllVolunteerTraining == False,
-                                   ((Event.program == None) |
+                                   ((Event.program_id == 9) |
```
Comment (AndersonStettner) [inline app/logic/events.py]: Like with Bonner events above, you can change `Event.program_id == 9` to `Event.isCeltsSponsored`

File: app/controllers/events/email.py [hunk 10]
```diff
@@ -32,7 +32,7 @@ def retrieveSenderList(eventId):
     senderOptions = [["CELTS (celts@berea.edu)", "Celts"]]
 
     event = Event.get_by_id(eventId)
-    if event.program_id:
+    if event.program_id != 9:
```
Comment (AndersonStettner) [inline app/controllers/events/email.py]: No magic number.

File: app/controllers/admin/routes.py [hunk 11]
```diff
@@ -48,17 +48,19 @@ def switchUser():
...
+                                celtsSponsoredProgram = celtsSponsoredProgram,
                                 templates=visibleTemplates)
     else:
         abort(403)
 
 
 @admin_bp.route('/eventTemplates/<templateid>/create', methods=['GET','POST'])
 @admin_bp.route('/eventTemplates/<templateid>/<programid>/create', methods=['GET','POST'])
-def createEvent(templateid, programid=None):
+def createEvent(templateid, programid=9):
```
Comment (AndersonStettner) [inline app/controllers/admin/routes.py]: No magic number.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

File: app/templates/events/event_list.html [hunk 12]
```diff
@@ -86,10 +86,10 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
             <tr class="">
           {% endif %}
             {% if type != "bonner" and type != "studentLed" %}
-              {% if event.program%}
-                <td class="text-dark">{{event.program.programName}}</td>
-              {% else %}
+              {% if event.isAllVolunteerTraining %}
```
Comment (BrianRamsay) [inline app/templates/events/event_list.html]: We don't really need to have this condition there anymore, because it has a program name

File: app/controllers/admin/routes.py [hunk 13]
```diff
@@ -48,17 +48,18 @@ def switchUser():
 def templateSelect():
     if g.current_user.isCeltsAdmin or g.current_user.isCeltsStudentStaff:
         allprograms = getAllowedPrograms(g.current_user)
+        celtsSponsoredProgram = Program.get(Program.programName == 'CELTS-Sponsored Event')
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: There is a database flag for this now. Don't select by program name

File: app/controllers/events/email.py [hunk 14]
```diff
@@ -32,7 +32,7 @@ def retrieveSenderList(eventId):
     senderOptions = [["CELTS (celts@berea.edu)", "Celts"]]
 
     event = Event.get_by_id(eventId)
-    if event.program_id:
+    if not event.program.isOtherCeltsSponsored:
```
Comment (BrianRamsay) [inline app/controllers/events/email.py]: Why are we treating Other CELTS Sponsored differently?

File: app/logic/transcript.py (hunk 8 above)
Comment (BrianRamsay) [inline app/logic/transcript.py]: No magic numbers. Use the database field

File: app/logic/transcript.py [hunk 15]
```diff
@@ -33,10 +33,11 @@ def getProgramTranscript(username):
     """
     # Add up hours earned in a term for each program they've participated in
     
+    OTHERCELTSSPONSORED = 9
     EventData = (Event
         .select(Event, fn.SUM(EventParticipant.hoursEarned).alias("hoursEarned"))
         .join(EventParticipant)
-        .where(EventParticipant.user == username)
+        .where(EventParticipant.user == username, Event.program != OTHERCELTSSPONSORED)
```
Comment (BrianRamsay) [inline app/logic/transcript.py]: Extra strange, considering we have a database field for this now.

File: app/logic/userManagement.py [hunk 16]
```diff
@@ -53,7 +53,7 @@ def changeProgramInfo(newProgramName, newContactEmail, newContactName, newLocati
 def getAllowedPrograms(currentUser):
     """Returns a list of all visible programs depending on who the current user is."""
     if currentUser.isCeltsAdmin:
-        return Program.select().order_by(Program.programName)
+        return Program.select().where(Program.isOtherCeltsSponsored == False).order_by(Program.programName)
```
Comment (BrianRamsay) [inline app/logic/userManagement.py]: Why are we excluding these programs? The previous query didn't restrict by program at all.

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

**Review by BrianRamsay — APPROVED** [review]
